### PR TITLE
AArch64: Enable Recompilation

### PR DESCRIPTION
**Pull request by knn-k**
This commit enables recompilation for AArch64.
Profiling is not supported yet.

Signed-off-by: KONNO Kazuhiro <konno@jp.ibm.com>

### CODE REVIEW

**Comment by knn-k** [comment]
@genie-omr build aarch64

**Comment by knn-k** [comment]
https://github.com/eclipse/openj9/pull/9847 ran sanity.functional tests depending on this PR, and they all passed.

**Comment by knn-k** [comment]
Build failed because UNB machines are offline today.
```
[2020-06-12T00:24:17.397Z] Still waiting to schedule task
[2020-06-12T00:24:17.399Z] All nodes of label ‘Linux&&x86&&"compile:aarch64"’ are offline
[2020-06-12T08:24:02.037Z] Cancelling nested steps due to timeout
```

**Comment by 0xdaryl** [comment]
@genie-omr build aarch64

File: compiler/aarch64/codegen/OMRCodeGenerator.cpp [hunk 1]
```diff
@@ -311,6 +329,15 @@ OMR::ARM64::CodeGenerator::doBinaryEncoding()
       }
 
    self()->getLinkage()->performPostBinaryEncoding();
+
+   if (data.recomp != NULL && data.recomp->couldBeCompiledAgain())
+      {
+      J9::PrivateLinkage::LinkageInfo *lkInfo = J9::PrivateLinkage::LinkageInfo::get(self()->getCodeStart());
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/OMRCodeGenerator.cpp:335]: You can't add J9-specific code in OMR.  Given that this code deals with private linkage, perhaps it should move into the OpenJ9 implementation of `performPostBinaryEncoding`, which is conveniently called just before this.

**Comment by knn-k** [comment]
@genie-omr build aarch64

**Comment by knn-k** [comment]
@genie-omr build aarch64

File: compiler/aarch64/codegen/OMRCodeGenerator.cpp [hunk 2]
```diff
@@ -236,11 +239,27 @@ OMR::ARM64::CodeGenerator::doBinaryEncoding()
...
+
+   while (data.cursorInstruction && data.cursorInstruction->getOpCodeValue() != TR::InstOpCode::proc)
+      {
+      data.estimate = data.cursorInstruction->estimateBinaryLength(data.estimate);
+      data.cursorInstruction = data.cursorInstruction->getNext();
+      }
+
+   tempInstruction = data.cursorInstruction;
+
+   if (data.recomp != NULL && !data.recomp->useSampling())
```
Comment (knn-k) [inline compiler/aarch64/codegen/OMRCodeGenerator.cpp:256]: This part also needs to be moved out of OMR.  `useSampling()` is defined in J9Recompilation.hpp.

**Comment by knn-k** [comment]
@genie-omr build aarch64

**Comment by knn-k** [comment]
This commit depends on https://github.com/eclipse/openj9/pull/9882

File: compiler/aarch64/codegen/OMRCodeGenerator.cpp (hunk 1 above)
Comment (knn-k) [reply compiler/aarch64/codegen/OMRCodeGenerator.cpp]: Moved the code to `performPostBinaryEncoding()` in the OpenJ9 side.

File: compiler/aarch64/codegen/OMRCodeGenerator.cpp (hunk 2 above)
Comment (knn-k) [reply compiler/aarch64/codegen/OMRCodeGenerator.cpp]: Moved the call to `useSampling()` to `TR_ARM64Recompilation::generatePrologue()`.

**Pull request merged**